artiq/dashboard/moninj/device_manager.py

```diff
@@ -0,0 +1,319 @@
+import asyncio
+import logging
+import typing
+from typing import Optional
+
+from PyQt5.QtWidgets import QMessageBox
+from sipyco.pc_rpc import AsyncioClient
+from sipyco.sync_struct import Subscriber, Notifier
+
+from artiq.coredevice.comm_moninj import TTLOverride, TTLProbe, ICommMonInj, CommMonInj
+from artiq.dashboard.moninj.util import setup_from_ddb
+from artiq.dashboard.moninj.widgets.dac import DACWidget
+from artiq.dashboard.moninj.widgets.dds import DDSWidget
+from artiq.dashboard.moninj.widgets.ttl import TTLWidget
+from artiq.tools import logger
+
+if typing.TYPE_CHECKING:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+class DeviceManagerProxy(ICommMonInj):
+    def __init__(self, notify_port: int):
+        self.notify = Notifier(dict({"monitor": dict(), "injection_status": dict(), "disconnect": []}))
+        self.ddb = dict()
+        self.notify_port = notify_port
+        self.core_connection = CommMonInj(self.on_monitor, self.on_injection_status,
+                                          self.on_disconnect)
+        self.ddb_notify = Subscriber(notifier_name="devices",
+                                     notify_cb=self.on_notify,
+                                     target_builder=self.on_notify_build)
+
+    async def on_notify(self, mod):
+        if mod["action"] != "init":
+            await self.reconnect()
+
+    async def on_notify_build(self, init):
+        self.ddb = init
+        await self.connect_coredev()
+        return self.ddb
+
+    def on_monitor(self, channel: int, probe: int, value: int):
+        if channel not in self.notify["monitor"].raw_view:
+            self.notify["monitor"][channel] = dict()
+        self.notify["monitor"][channel][probe] = value
+
+    def on_injection_status(self, channel: int, override: int, value: int):
+        if channel not in self.notify["injection_status"].raw_view:
+            self.notify["injection_status"][channel] = dict()
+        self.notify["injection_status"][channel][override] = value
+
+    def on_disconnect(self):
+        self.notify["disconnect"].append(None)
+
+    async def reconnect(self):
+        await self.stop()
+        await self.connect()
+
+    async def stop(self):
+        try:
+            await self.ddb_notify.close()
+            await self.core_connection.close()
+        except:
+            pass
+
+    async def connect(self):
+        await self.ddb_notify.connect("localhost", self.notify_port)
```

Comment:
localhost?

Reply:
it is for master to receive update for device database update to restart the core device proxy, and far as I know, I couldn't find anyway to run an in-memory subscriber `sipyco.sync_struct`to listen to the device database update publisher. In the mean time it should just work

Reply:
invalid now, changes are made to the proxy module, so this is no longer running on master directly and cannot simply assume localhost

Reply:
Invalid not just now, the master isn't always on localhost either.